core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java

```diff
@@ -81,6 +87,35 @@ public static EnumerableCorrelate create(
         traitSet, left, right, correlationId, requiredColumns, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    final RelCollation collation = required.getCollation();
+    if (collation == null || collation == RelCollations.EMPTY) {
+      return null;
+    }
+
+    // EnumerableCorrelate traits passdown shall only pass through collation to left input.
+    // This is because for EnumerableCorrelate always uses left input as the outer loop,
+    // thus only left input can preserve ordering.
+
+    for (RelFieldCollation relFieldCollation : collation.getFieldCollations()) {
+      // If field collation belongs to right input: bail out.
+      if (relFieldCollation.getFieldIndex() >= getLeft().getRowType().getFieldCount()) {
+        return null;
+      }
+    }
+
+    final RelTraitSet passThroughTraitSet = traitSet.replace(collation);
+    return Pair.of(passThroughTraitSet,
+        ImmutableList.of(
+            passThroughTraitSet,
+            passThroughTraitSet.replace(RelCollations.EMPTY)));
+  }
+
+  @Override public DeriveMode getDeriveMode() {
+    return DeriveMode.LEFT_FIRST;
+  }
```

Comment:
`deriveTraits` is still missing. If the left child can deliver some collation, the correlate should be able to deliver the same collation.

Reply:
Yes. Any collation can be derived from left input. Optimizer can decide later if derived collations are useful.

Reply:
`deriveTraits` method added, not sure if the implementation is correct though, please take a look.